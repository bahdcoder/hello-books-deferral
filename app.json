{
  "name": "Hello Books",
  "description": "A system that allows efficient, cost-effective management of Books for Libraries."
}
