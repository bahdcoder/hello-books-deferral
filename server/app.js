import express from 'express';
import consola from 'consola';
import bodyParser from 'body-parser';
import swagger from 'swagger-ui-express';
import Knex from 'knex';
import { createServer } from 'http';
import { Model } from 'objection';
import swaggerDocument from '@/swagger.json';
import Routes from '@routes/v1';
import config from '@config';

import 'express-jsend';

export const app = express();

export const databaseConnection = Knex(config.database[config.server.env]);

Model.knex(databaseConnection);

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: true }));

Routes(app);

app.get('/', (req, res) => {
    res.status(200).jsend({ message: 'Hello Books Deferral' });
});
/**
 *  setup Swagger
 */
app.use(
    '/api-docs',
    swagger.serve,
    swagger.setup(swaggerDocument, { explorer: true })
);

export const server = createServer(app);

export default { app, databaseConnection };
