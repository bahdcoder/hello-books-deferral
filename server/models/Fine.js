import { Model } from 'objection';

class Fine extends Model {
    static tableName = 'fines';
}

export default Fine;
