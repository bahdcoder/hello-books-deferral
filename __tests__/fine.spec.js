import Fine from '@models/Fine';
import supertest from 'supertest';
import { app, databaseConnection } from '@server/app';
import {
    getUser,
    createUser,
    createFine,
    getFine,
    getToken
} from '@tests/utils/helpers';

const server = () => supertest(app);
let adminToken;

describe.skip('FINES API ENDPOINTS', () => {
    beforeAll(async () => {
        await databaseConnection.migrate.latest();
        await databaseConnection('users').truncate();
        await databaseConnection('fines').truncate();
        const thisUser = await createUser(getUser());
        const admin = getUser();
        admin.role = 'admin';
        const theAdmin = await createUser(admin);
        await createFine(thisUser.id);
        adminToken = getToken({ id: theAdmin.id, email: theAdmin.email });
    });

    afterAll(async () => {
        await databaseConnection('fines').truncate();
        await databaseConnection('users').truncate();
        server.close();
    });

    describe('GET FINES api/v1/fines', () => {
        it('should get all fines for authenticated user', async () => {
            const user = await createUser(getUser());
            await createFine(user.id);
            const userToken = getToken({ id: user.id, email: user.email });
            const { status, body } = await server()
                .get('/api/v1/fines')
                .set('x-access-token', userToken);
            expect(status).toBe(200);
            expect(body.data.length).toBe(1);
            expect(body.data[0].user_id).toBe(user.id);
        });
    });

    describe('POST CREATE FINE api/v1/fines', () => {
        it('should not create fine if type and description field is empty', async () => {
            const { status, body } = await server()
                .post('/api/v1/fines/1')
                .set('x-access-token', adminToken)
                .send({
                    ...getFine(1),
                    type: '',
                    description: ''
                });

            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should not create fine if userId param is not a number', async () => {
            const { status, body } = await server()
                .get('/api/v1/fines/bag')
                .set('x-access-token', adminToken)
                .send(getFine(1));
            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should not create fine if userId does not exists', async () => {
            const { status, body } = await server()
                .post('/api/v1/fines/1545678')
                .set('x-access-token', adminToken)
                .send(getFine(1));
            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should create a fine with valid inputs', async () => {
            const { status, body } = await server()
                .post('/api/v1/fines/1')
                .set('x-access-token', adminToken)
                .send(getFine(1));

            expect(status).toBe(201);
            expect(body.data.userFine.user_id).toBe('1');
        });
    });

    describe('GET FINE api/v1/fines', () => {
        it('should not get fine if fineId param is not a number', async () => {
            const { status, body } = await server()
                .get('/api/v1/fines/5ku')
                .set('x-access-token', adminToken);
            expect(status).toBe(422);
            expect(body).toMatchSnapshot();
        });

        it('should return a fine', async () => {
            const { status, body } = await server()
                .get('/api/v1/fines/1')
                .set('x-access-token', adminToken);

            expect(status).toBe(200);
            expect(body.data.amount).toBe(50000);
        });
    });
});
