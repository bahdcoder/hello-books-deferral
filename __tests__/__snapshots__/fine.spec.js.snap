// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FINES API ENDPOINTS DELETE FINES api/v1/fines should delete fines with id param valid 1`] = `
Object {
  "data": Object {
    "message": "Fine succesfully deleted",
  },
  "status": "success",
}
`;

exports[`FINES API ENDPOINTS GET FINE api/v1/fines should not get fine if fineId param is not a number 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "fineId",
      "message": "Fine Id does not exist",
      "validation": "itExists",
    },
  ],
  "status": "error",
}
`;

exports[`FINES API ENDPOINTS POST CREATE FINE api/v1/fines should not create fine if type and description field is empty 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "type",
      "message": "type is required to create a fine",
      "validation": "required",
    },
    Object {
      "field": "description",
      "message": "description is required to create a fine",
      "validation": "required",
    },
  ],
  "status": "error",
}
`;

exports[`FINES API ENDPOINTS POST CREATE FINE api/v1/fines should not create fine if userId does not exists 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "userId",
      "message": "User's Id does not exist",
      "validation": "itExists",
    },
  ],
  "status": "error",
}
`;

exports[`FINES API ENDPOINTS POST CREATE FINE api/v1/fines should not create fine if userId param is not a number 1`] = `
Object {
  "code": "ValidationFailed",
  "message": Array [
    Object {
      "field": "fineId",
      "message": "fineId is expected to be a an integer",
      "validation": "number",
    },
    Object {
      "field": "fineId",
      "message": "select \\"fines\\".* from \\"fines\\" where \\"id\\" = $1 - invalid input syntax for integer: \\"bag\\"",
      "validation": "ENGINE_EXCEPTION",
    },
  ],
  "status": "error",
}
`;
